File: packages/contracts/src/promotion-run.ts
import { z } from 'zod';
import { ActorTypeSchema } from './audit.js';

export const PromotionStatusSchema = z.enum([
  'pending',
  'active',
  'paused',
  'completed',
  'aborted',
]);

export const ActorSchema = z.object({
  actorType: ActorTypeSchema,
  actorId: z.string(),
  displayName: z.string().optional(),
});

export const PromotionRunCreateInputSchema = z.object({
  pipelineId: z.string(),
  flagKey: z.string(),
  actor: ActorSchema,
});

export const PersistRunStateInputSchema = z.object({
  promotionRunId: z.string(),
  status: PromotionStatusSchema,
  currentStageIndex: z.number().int().optional(),
  pauseReason: z.string().optional(),
});

export type PromotionStatus = z.infer<typeof PromotionStatusSchema>;
export type Actor = z.infer<typeof ActorSchema>;
export type PromotionRunCreateInput = z.infer<typeof PromotionRunCreateInputSchema>;
export type PersistRunStateInput = z.infer<typeof PersistRunStateInputSchema>;

File: kotlin/modules/contracts/src/main/kotlin/com/ffpromo/contracts/PromotionRun.kt
package com.ffpromo.contracts

import kotlinx.serialization.Serializable

@Serializable
data class PromotionRunCreateInput(
    val pipelineId: String,
    val flagKey: String,
    val actor: Actor,
)

@Serializable
data class PersistRunStateInput(
    val promotionRunId: String,
    val status: PromotionStatus,
    val currentStageIndex: Int? = null,
    val pauseReason: String? = null,
)
